### DISCUSSION TITLE
custom action and feature

### DISCUSSION THREAD

**Discussion by pas059** [wixtoolset/issues#7633] [Questions] [answered]
Hello,
I want to add a custom action to my installer for installing a driver. But i don't want to install this driver if the user doesn't select the corresponding feature (having Id=Driver_USBXpress_67_Feature). To do this i added the following code to my Product.wxs file:

`<CustomAction Id="InstallUsbXpress" BinaryRef="WixCA_UsbxpressDriver.CA" DllEntry="CA_InstallUsbXpress" Execute="deferred" Return="check" Impersonate="no" />
`
`<Property Id="InstallUsbXpress" Value="DriverName=siusbxp.inf" />`
`<Binary Id="WixCA_UsbxpressDriver.CA" SourceFile="..\WixCA_UsbxpressDriver\bin\Release\net472\WixCA_UsbxpressDriver.CA.dll" />`
`<InstallExecuteSequence>`
`   <Custom Action="InstallUsbXpress" After="InstallFiles" Condition="NOT Installed AND !Driver_USBXpress_67_Feature=3" />`
`</InstallExecuteSequence>`

Someone can tell if this code is correct, especially the 'Custom' element?

**Comment by pas059** [comment]
Hi,
i did not find the solution. It's like the CA does not start, but i not sure about this.
Someone can give link(s) on how to use a CA in WiX 4, and how to debug it.

**Comment by pas059** [comment, marked as answer]
Hi,
here the following of my experience....

1. i wasn't aware that the CA dll has to be a strong name assembly, which creates many problems after trying to install (exception in rundll32.exe? , severe difficulties to uninstall,...)
2. the feature test was incorrect, below a one that works

```
<InstallExecuteSequence>
	<Custom Action="InstallUsbXpress" 
			After="InstallFiles"
			Condition="NOT Installed AND &amp;Driver_USBXpress_67_Feature=3"
	/>
</InstallExecuteSequence>
```
So, now my custom action starts, but fails because it tries to launch a x64 process (C:\Windows\System32\pnpuil.exe) although the generated CA.dll is a x86 dll. I tried other dll generations but without success until now. But i will open another topic for that,  except if some one has an idea.

**Reply by robmen** [reply]
64-bit folders are generally hidden from 32-bit processes.

**Reply by pas059** [reply]
Hi,
i had posted another topic [here](https://github.com/orgs/wixtoolset/discussions/7656) to explains what happens with more details.
